src/zkevm_specs/evm/execution/end_block.py

```diff
@@ -4,69 +4,102 @@
+#
+# We prove (A) by constraining the transition rule that after an EndBlock
 # state, only an EndBlock state can follow.
 #
-# We prove (2) and (3) by proving that at least `MAX_COUNT - evm_circuit_count`
-# padding elements exist in the rw table and tx table.  To achieve this, we do
-# padding at the beginning of both tables with sequential indexes at
-# `rw_counter` (in rw_table) or `tx_id` (in tx_table).
+# We prove (B) by proving that at least `MAX_RWS - total_rws` padding elements
+# exist in the rw table.  To achieve this, we do padding at the beginning of
+# the RwTable with a sequential index `rw_counter`.  The RwTable constraints
+# that padding rows have sequential indexes, so by doing 2 lookups with
+# rw_counter = {a, b} we know there are at least b-a padding rows.
+#
+# We prove (C) by showing that the transaction ID after the last one processed
+# corresponds to a padding tx in the TxTable, which enforces that once a
+# padding txs appears in the table, the rest will also be padding.  This way,
+# we know that if a padding tx with tx_id exist, then at most there are tx_id-1
+# non-padding txs.  In case we have exhausted the TxTable with txs, we won't
+# have any padding txs; so we skip the lookup.
 #
-# We prove (4) in the circuit implementation by constraining that the last
+# We prove (D) in the circuit implementation by constraining that the last
 # execution step at the end of the EVM circuit is an EndBlock.  When the number
 # of steps is less than the EVM circuit height, we pad at the end with
 # EndBlock.  This will require the EndBlock to have height = 1 in the circuit,
 # which can be achieved after reducing the number of cells used in the state
 # selector.
 
 # Count the max number of txs that the TxTable can hold by counting rows of
-# fixed fields + padding rows in the fixed fields section.
+# type CallerAddress.
 def get_tx_table_max_txs(table: Set[TxTableRow]) -> int:
-    fixed_field_count = 0
-    for row in table:
-        if (row.field_tag != TxContextFieldTag.CallData) or (
-            row.field_tag == TxContextFieldTag.Pad and row.tx_id != 0
-        ):
-            fixed_field_count += 1
-    return fixed_field_count // TxContextFieldTag.TxSignHash
+    return len([row for row in table if row.field_tag == TxContextFieldTag.CallerAddress])
 
 
 def end_block(instruction: Instruction):
+    max_txs = get_tx_table_max_txs(instruction.tables.tx_table)
+    max_rws = len(instruction.tables.rw_table)
+    total_txs = FQ(
+        len(
+            [
+                tx_row
+                for tx_row in instruction.tables.tx_table
+                if tx_row.field_tag == TxContextFieldTag.CallerAddress
+                and tx_row.value.expr() != FQ(0)
+            ]
+        )
+    )
+    last_rw_counter = instruction.curr.rw_counter - 1
+
+    total_rws = last_rw_counter
+    if total_txs.n > 0:
+        # If the block is not empty, we will do a call_context lookup
+        total_rws = last_rw_counter + 1
+
```

Comment:
It seems a little bit difficult to understand when `total_rws` depends on `total_txs`, could we try to define `total_rws` in this way?

```python
is_empty_block = is_zero(instruction.curr.rw_counter)
if is_empty_block:
    total_rws = 0
else:
    total_rws = instruction.curr.rw_counter # extra 1 from tx_id lookup
```

Note currently this makes `total_rws` to be always `instruction.curr.rw_counter`, but in the future we might have more lookup in `EndBlock` (for rewarding block proposer), then this if-else is necessary.

Reply:
Thanks for the suggestion, I've applied it in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/286/commits/21e6a522ce2e0fbcdfb78759c4526d108fbf94ba